### DISCUSSION TITLE
Developing OpenHantek, errors, fixes, source code

### DISCUSSION THREAD

**Discussion by nseam** [OpenHantek/OpenHantek6022#376] [General]
Hi!

Just a little question - did you guys ever experienced `vector's iterator dereference out of range` error in `GraphGenerator::generateGraphsTYvoltage( PPresult *result )` in **debug mode**? I just can't use the app compiled in debug mode. In release mode of course the exceptions are silences so app works correctly. Currently I don't know what's the issue. Leaving this unresolved could lead to some unexpected problems with graphs' rendering.

![image](https://github.com/OpenHantek/OpenHantek6022/assets/1587690/6fe12bb0-86f4-4904-af53-6687ece1d21c)

![image](https://github.com/OpenHantek/OpenHantek6022/assets/1587690/7bfefdd7-29d6-4093-9086-4a4b5024fa33)

Thanks!

**Comment by Ho-Ro** [comment]
Never saw it, but I think it is caused by this [code part](https://github.com/OpenHantek/OpenHantek6022/blob/8bc0b8b0f99f789283a3ae95f50baec7894b9007/openhantek/src/post/graphgenerator.cpp#L146-L150):
```c++
        for ( unsigned int position = unsigned( leftmostPosition ); position < dotsOnScreen && sampleIterator < sampleEnd;
              ++position ) {
            double x = double( MARGIN_LEFT + position * horizontalFactor );
            double y_1 = *sampleIterator++ / gain + offset;
            double y = *sampleIterator / gain + offset;
```
`sampleEnd` is either `auto sampleEnd = sampleValues.samples.cend();` or `sampleEnd = resample.cend();`, so it points past the last element, and the last access `double y = *sampleIterator / gain + offset;` is one off.

I will fix it with `... && sampleIterator < sampleEnd - 1;` in the 1st line.

Thank you for reporting!

**Reply by nseam** [reply]
I would probably check it today's evening about that error, because it looked easy to fix. It's just strange that nobody reported it before.

**Reply by Ho-Ro** [reply]
I put a check for `sampleIterator >= sampleEnd` after the loop and it was triggered without my fix - no fail after the fix.
Fixed in 3.3.3-rc2
Thx, please keep on testing :)

**Comment by nseam** [comment]
I have also some trivial issue. When I'm running app through Visual Studio then I have all the icons in the toolbar. When ran via .exe then there's no icons there. I tried to copy folders/.res/.* files from there to there and nothing helped. Maybe you know what could be the issue? It is visually unpleasant and I wanted to use debug version of the app so I can debug it on PC or laptop when some error will occur.

**Reply by Ho-Ro** [reply]
No clue, I edit in `qtcreator` on Linux Debian stable, but use `cmake` to prepare the build and `make` to build the binary.
